onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,622 @@
+    std::vector<NodeArg*>& outputs = node->MutableOutputDefs();
+    std::unordered_set<NodeArg*> require_type_change;
+    for (auto iter = outputs.begin(); iter != outputs.end() && all_float_outputs_have_casts; ++iter) {
+      NodeArg* output = *iter;
+      if (output->TypeAsProto()->tensor_type().elem_type() != TensorProto::FLOAT) {
+        continue;
+      }
+      has_float_outputs = true;
+      std::vector<Node*> consumers = graph.GetMutableConsumerNodes(output->Name());
+      for (auto node_iter = consumers.begin(); node_iter != consumers.end() && all_float_outputs_have_casts; ++node_iter) {
+        Node* consumer = *node_iter;
+        if (consumer && consumer->OpType() == "Cast") {
+          const NodeAttributes& attributes = consumer->GetAttributes();
+          ORT_ENFORCE(attributes.find("to") != attributes.end());
+          if (attributes.at("to").i() == static_cast<int64_t> (TensorProto::FLOAT16)) {
+            casts.push_back(consumer);
+            continue;
+          }
+        }
+        all_float_outputs_have_casts = false;
+      }
+      require_type_change.insert(output);
+    }
+    if (has_float_outputs && all_float_outputs_have_casts && casts.size() > 0 ) {
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Removed Cast nodes "
+                       << GatherNames<std::vector<Node*>>(casts)
+                       << " feeding the same compute node " << node->Name();
+      RemoveCastNodes(graph, casts);
+      std::unordered_set<NodeArg*> node_args;
+      for (NodeArg* input : node->MutableInputDefs()) {
+        if (input->TypeAsProto()->tensor_type().elem_type() == TensorProto::FLOAT) {
+          node_args.insert(input);
+        }
+      }
+      InsertCastNodes(graph, node_args, true);
+      ChangeTypeToFP16(require_type_change, logger);
+      VLOGS(logger, 1) << "PropagateCastsFromOutputsToInputs: Inserted Cast node to " << GatherNames(node_args);
+      modified = true;
+    }
+  }
+  return modified;
+}
+
+// Expand FP16 compute regions on the graph by example float16 compute nodes,
+// propagating float32 Cast operation down the graph and propagating float16
+// Cast operations up the graph. The following functions are performed
+// 1. Fuse subgraphs
+// 2. Propagate fp32 casts forwards
+// 3. Propagate fp16 casts back
+Status PropagateCastOps::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  // First apply the transformation to the subgraphs.
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto node_index : node_topology_list) {
+    auto* node_ptr = graph.GetNode(node_index);
+    if (nullptr == node_ptr)
+      continue;  // node was removed
+
+    auto& node = *node_ptr;
+
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+  }
+
+  bool local_modified = false;
+  do {
+    if (local_modified) {
+      graph.Resolve();
```

Comment:
do we need to call graph.resolve() between different propagation functions? 

I am wondering if there will be any interaction between the propagation functions that could mess up the graph

Reply:
I added remove_nodes to collect the nodes to be removed and only by the end of the do-while-loop I am removing the nodes and also making sure to skip removed nodes.